Source/DafnyLanguageServer/Workspace/TelemetryPublisher.cs

```diff
@@ -0,0 +1,27 @@
+﻿using Microsoft.Dafny.LanguageServer.Util;
+using Microsoft.Dafny.LanguageServer.Workspace.Notifications;
+using OmniSharp.Extensions.LanguageServer.Protocol.Document;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using OmniSharp.Extensions.LanguageServer.Protocol.Server;
+using System.Collections.Generic;
+using System.Linq;
+using OmniSharp.Extensions.LanguageServer.Protocol.Window;
+
+namespace Microsoft.Dafny.LanguageServer.Workspace {
+  public class TelemetryPublisher : ITelemetryPublisher {
+    private readonly ILanguageServerFacade languageServer;
+
+    public TelemetryPublisher(ILanguageServerFacade languageServer) {
+      this.languageServer = languageServer;
+    }
+
+    void ITelemetryPublisher.PublishTelemetry(ITelemetryPublisher.TelemetryEventKind kind, object? payload) {
+      languageServer.Window.SendTelemetryEvent(new TelemetryEventParams {
+        ExtensionData = new Dictionary<string, object> {
+          {"kind", kind.ToString()},
+          {"data", payload ?? new Dictionary<string, object>()}
```

Comment:
Maybe rename `data` to `payload` to reflect the argument name.
And personal preference: I'd leave the `data` property of the serialized object if it's empty rather than falling back to an empty dictionary.

Reply:
Rename done.  I don't know how to leave the payload empty: annoyingly `TelemetryEventParams` expects a `Dictionary<string, object>`, not a `Dictionary<string, object?>`, so I can't put `null` there.